propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb

```diff
@@ -0,0 +1,117 @@
+        # @param [Context] context The context to be updated with extracted
+        #   context
+        # @param [optional Callable] getter An optional callable that takes a
+        #   carrier and a key and returns the value associated with the key. If
+        #   omitted the default getter will be used which expects the carrier to
+        #   respond to [] and []=.
+        # @yield [Carrier, String] if an optional getter is provided, extract
+        #   will yield the carrier and the header key to the getter.
+        # @return [Context] Updated context with active span derived from the
+        #   header, or the original context if parsing fails.
+        def extract(carrier, context, getter = nil)
+          getter ||= @default_getter
+          header = getter.get(carrier, IDENTITY_KEY)
+          return context unless (match = header.match(TRACE_SPAN_IDENTITY_REGEX))
+          return context if match['trace_id'] =~ /^0+$/
+          return context if match['span_id'] =~ /^0+$/
+
+          sampling_flags = match['sampling_flags'].to_i
+          span = build_span(match, sampling_flags)
+          context = Jaeger.context_with_debug(context) if sampling_flags & DEBUG_FLAG_BIT != 0
+          context = set_baggage(carrier, context, getter)
+          Trace.context_with_span(span, parent_context: context)
+        end
+
+        private
+
+        def build_span(match, sampling_flags)
+          trace_id = to_trace_id(match['trace_id'])
+          span_context = Trace::SpanContext.new(
+            trace_id: trace_id,
+            span_id: to_span_id(match['span_id']),
+            trace_flags: to_trace_flags(sampling_flags),
+            remote: true
+          )
+          Trace::Span.new(span_context: span_context)
+        end
+
+        def set_baggage(carrier, context, getter)
+          baggage_key_prefix = 'uberctx-'
+          getter.keys(carrier).each do |carrier_key|
+            baggage_key = carrier_key.start_with?(baggage_key_prefix) && carrier_key[baggage_key_prefix.length..-1]
+            next unless baggage_key
+
+            raw_value = getter.get(carrier, carrier_key)
+            value = URI.decode_www_form_component(raw_value)
+            context = OpenTelemetry.baggage.set_value(
+              baggage_key, value, context: context
+            )
+          end
+          context
+        end
+
+        def to_trace_flags(sampling_flags)
+          if (sampling_flags & SAMPLED_FLAG_BIT) != 0
+            Trace::TraceFlags::SAMPLED
+          else
+            Trace::TraceFlags::DEFAULT
+          end
+        end
+
+        def to_span_id(span_id_str)
+          zero_pad_count = (16 - span_id_str.length) % 16
+          to_zero_padded_id(span_id_str, zero_pad_count)
+        end
+
+        def to_trace_id(trace_id_str)
+          zero_pad_count = (16 - (trace_id_str.length % 16)) % 16
+          to_zero_padded_id(trace_id_str, zero_pad_count)
```

Comment:
Similarly, I think this can be:
```suggestion
         trace_id.rjust(32, '0')
```

Reply:
The current implementation emits 64-bit trace IDs when it receives 64-bit trace IDs; using `rjust(32, '0')` would force them to zero-padded 128-bit trace IDs. Both would seem to match the Jaeger spec, but given that the spec says "Clients in some languages support 128-bit, migration pending", that would imply to me that some clients **do not** support 128-bit. Meaning that it might be marginally safer to stick with 64-bit trace IDs in case there are legacy clients downstream.

Or is that gold-plating? Happy to just change it too.

Reply:
The [opentelemetry spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spancontext) says:

> TraceId A valid trace identifier is a 16-byte array with at least one non-zero byte.

So, I think when extracting we can safely zero pad to 32 hex chars.

Reply:
Sounds good, I'll zero-pad to 32 hex chars.